### ISSUE TITLE
Warning: Trying to access array offset on value of type null in Mview/View/Subscription.php

### ISSUE THREAD

**Issue by Nuranto** [magento/magento2#35735]
### Preconditions and environment

- Magento version 2.4.4
- PHP8.1

### Steps to reproduce

1. Create a custom Module
2. Configure an indexer
3. Configure an mview subscription
4. Enable module, purge cache, setup upgrade
5. Disable module, purge cache, setup upgrade

### Expected result

Setup upgrade finishes without error

### Actual result

```
Running M2 DB upgrade ...
Cache types config flushed successfully
Cache cleared successfully
Updating modules:
 Warning: Trying to access array offset on value of type null in /var/www/html/vendor/magento/framework/Mview/View/Subscription.php on line 307

```

### Additional information

Could be dangerous in production rollback scenarios.

### Triage and priority

- [ ] Severity: **S0** _- Affects critical data or functionality and leaves users without workaround._
- [ ] Severity: **S1** _- Affects critical data or functionality and forces users to employ a workaround._
- [X] Severity: **S2** _- Affects non-critical data or functionality and forces users to employ a workaround._
- [ ] Severity: **S3** _- Affects non-critical data or functionality and does not force users to employ a workaround._
- [ ] Severity: **S4** _- Affects aesthetics, professional look and feel, “quality” or “usability”._

**Comment by engcom-Echo** [comment]
Hello @Nuranto,

Currently I have below installed versions on my local machine and the issue is not reproducible.
Magento 2.4-develop
PHP 8.1.7

In order to reproduce the issue followed below steps.

1. Created a custom Module ‘Magento_CustomModule’
2. Created an Indexer ‘RH Custom Indexer’
 
![MicrosoftTeams-image](https://user-images.githubusercontent.com/51681435/180421738-f29e51da-3974-4662-8aa5-73b37b17b0a8.png)

3. Created a mView subscription

![MicrosoftTeams-image (2)](https://user-images.githubusercontent.com/51681435/180421796-a8973c36-315e-4fbc-8d35-cb9cdbd63a7b.png)

4. Enabled the module
5. Clean cache and setup upgrade
6. Disabled the module 
7. Cleaned cache and setup upgrade

Setup:upgrade command is running successfully without error.  Hence, this is not reproducible.
Please let me know if I am missing anything in these mentioned steps to reproduce the issue.

Thanks

**Comment by Nuranto** [comment]
Indeed, I can't reproduce on a fresh installation as well.
And I can't find the missing step to reproduce. I'll close this for now, and will reopen if I find the glitch.

**Comment by GamesmenJordan** [comment]
Also encountered this error (2.4.3-p2), not sure what triggered it but it occurred during a standard deployment. 

Same codebase did not trigger the error when deploying to staging, re-running the production deployment with no changes completed successfully.

So seems like we have some kind of Schrodinger's bug.

**Comment by Hanhan1989** [comment]
Also encountered this error in Magento (2.4.5-p1), after updating **composer.json**.
I think it is due to the update of one of the following dependencies.

```

Updating dependencies
Lock file operations: 0 installs, 27 updates, 0 removals
  - Upgrading aws/aws-sdk-php (3.253.2 => 3.257.5)
  - Upgrading brick/varexporter (0.3.7 => 0.3.8)
  - Upgrading composer/ca-bundle (1.3.4 => 1.3.5)
  - Upgrading doctrine/annotations (1.14.1 => 1.14.2)
  - Upgrading fgrosse/phpasn1 (v2.4.0 => v2.5.0)
  - Upgrading laminas/laminas-captcha (2.15.0 => 2.16.0)
  - Upgrading laminas/laminas-db (2.16.2 => 2.16.3)
  - Upgrading laminas/laminas-di (3.11.0 => 3.12.0)
  - Upgrading laminas/laminas-eventmanager (3.9.0 => 3.10.0)
  - Upgrading laminas/laminas-mail (2.21.0 => 2.22.0)
  - Upgrading laminas/laminas-router (3.11.0 => 3.11.1)
  - Upgrading laminas/laminas-view (2.25.0 => 2.26.0)
  - Upgrading magento/magento-coding-standard (27 => 29)
  - Upgrading nikic/php-parser (v4.15.2 => v4.15.3)
  - Upgrading phpseclib/mcrypt_compat (2.0.3 => 2.0.4)
  - Upgrading phpseclib/phpseclib (3.0.17 => 3.0.18)
  - Upgrading phpstan/phpstan (1.9.3 => 1.9.14)
  - Upgrading ramsey/collection (1.2.2 => 1.3.0)
  - Upgrading rector/rector (0.13.10 => 0.14.8)
  - Upgrading symfony/css-selector (v6.2.0 => v6.2.3)
  - Upgrading symfony/dependency-injection (v5.4.16 => v5.4.17)
  - Upgrading symfony/finder (v5.4.11 => v5.4.17)
  - Upgrading symfony/http-foundation (v5.4.16 => v5.4.17)
  - Upgrading symfony/var-dumper (v5.4.14 => v5.4.17)
  - Upgrading webonyx/graphql-php (v14.11.8 => v14.11.9)
  - Upgrading wikimedia/less.php (v3.1.0 => v3.2.0)
Writing lock file



```

**Comment by dbashyal** [comment]
I also encountered this error in Magento (2.4.5-p1)
![image](https://user-images.githubusercontent.com/833986/215638168-8ad85e04-73e2-4878-95f9-fd93d79c844d.png)

2nd attempt: `s:upgrade --keep-generated` didn't throw error after running "composer install" again.

**Comment by peterjaap** [comment]
Also run into this on 2.4.6 (clean install) after changing the indexers from realtime to schedule.

```
php /data/web/releases/20230314194449/bin/magento setup:upgrade --keep-generated --no-interaction
Cache types config flushed successfully
Cache cleared successfully
Updating modules:
Warning: Trying to access array offset on value of type null in /data/web/releases/20230314194449/vendor/magento/framework/Mview/View/Subscription.php on line 344
```

Restarting the deployment job didn't help for me.

**Comment by edza345** [comment]
> Also run into this on 2.4.6 (clean install) after changing the indexers from realtime to schedule.
> 
> ```
> php /data/web/releases/20230314194449/bin/magento setup:upgrade --keep-generated --no-interaction
> Cache types config flushed successfully
> Cache cleared successfully
> Updating modules:
> Warning: Trying to access array offset on value of type null in /data/web/releases/20230314194449/vendor/magento/framework/Mview/View/Subscription.php on line 344
> ```
> 
> Restarting the deployment job didn't help for me.
> 
> When I dump out the mview ID with this in `Mview/View/Subscription.php`:
> 
> ```
>     private function getProcessor(): AdditionalColumnProcessorInterface
>     {  
>         print-r($this->getView()->getData());
> ```
> 
> I get;
> 
> ```
> Array
> (
>     [subscriptions] => Array
>         (
>             [cataloginventory_stock_item] => Array
>                 (
>                     [name] => cataloginventory_stock_item
>                     [column] => 
>                     [subscription_model] => 
>                 )
> 
>         )
> 
>     [view_id] => old_view
> )
> ```
> 
> This view_id value is set in the TriggerCleaner class;
> 
> https://github.com/magento/magento2/blob/176e388fb6168f76d1d8531b2375d72aaa00ba81/lib/internal/Magento/Framework/Mview/TriggerCleaner.php#L162
> 
> That does a query on `information_schema` database to see which triggers are currently there. When I run this query;
> 
> ```
> SELECT TRIGGER_NAME FROM information_schema.TRIGGERS WHERE TRIGGER_SCHEMA='my_magento_db' AND TRIGGER_NAME LIKE '%cataloginventory_stock_item%';
> ```
> 
> I get these results;
> 
> ```
> +----------------------------------------------+
> | TRIGGER_NAME                                 |
> +----------------------------------------------+
> | trg_cataloginventory_stock_item_after_insert |
> | trg_cataloginventory_stock_item_after_update |
> | trg_cataloginventory_stock_item_after_delete |
> +----------------------------------------------+
> ```
> 
> So I'm guessing it has something to do with triggers that aren't created but should, or vice-versa?
> 
> What helped me out for now is making sure Magento does not try to remove subscribers on the `old_view` trigger since apparently it doesn't have any anyway.
> 
> It can't remove subscribers when there's no processor and it can't create a processor because it doesn't have an action class.
> 
> So in `vendor/magento/framework/Mview/TriggerCleaner.php:90`, I've added an if statement;
> 
> ```
>             $view = $this->createViewByTableName($trigger['EVENT_OBJECT_TABLE']);
>             if ($view->getActionClass()) {
>                 $view->unsubscribe();
>                 $view->getState()->delete();
>             }
> ```
> 
> This makes my setup:upgrade pass. But I'm not entirely sure whether this won't have any side-effects down the line...

Thank you, this helped fixing this annoying upgrade issue. Did not go into details what was causing it, but I did create patch for easier application for someone who also having this issue

https://www.dropbox.com/s/9ml89ya9z2sx0f1/trigger_cleaner_fix.patch?dl=0
From magento root run it with: patch -p1 < clkweb_trigger_cleaner_fix.patch

**Comment by jonathanribas** [comment]
In our case it's happening because we have custom triggers on our database not plugged on `mview` system at all.

**Comment by jonatanrdsantos** [comment]
I had the same issue upgrading 2.4.5 to 2.4.6. The patch worked fine.

**Comment by danielraab** [comment]
The same error here but for the `catalog_product_super_link` subscription. The workaround is working for me too. But i wonder why this occurs now ? I see no changes in Magento/Framework/Mview between 2.4.5 and 2.4.6 ??

**Comment by jonatanrdsantos** [comment]
Hi all, I was able to run the upgrade by removing and installing the triggers as per: https://github.com/magento/magento2/issues/23023#issuecomment-497383606

**Comment by Davie82** [comment]
I had this issue when upgrading from 2.4.4 to 2.4.6.

When I dump out the mview id I also get:

```
Array
(
    [subscriptions] => Array
        (
            [cataloginventory_stock_item] => Array
                (
                    [name] => cataloginventory_stock_item
                    [column] => 
                    [subscription_model] => 
                )

        )

    [view_id] => old_view
)
```

I compared the triggers for `cataloginventory_stock_item` with a clean Magento 2.4.6. and found some differences.

Current:
<img width="983" alt="current-version" src="https://github.com/magento/magento2/assets/17762698/7b0ff295-72b8-46c4-8456-2aa33a198a1b">

Clean Magento:
<img width="916" alt="clean-magento" src="https://github.com/magento/magento2/assets/17762698/bd907eed-72d0-4a44-9a28-d8322a8047c8">

So, `stock_after_insert` and `stock_before_update` are not present in a clean Magento anymore.

I manually removed them and after that the error is not showing anymore when running setup:upgrade.

I'm not sure if removing these triggers will cause other problems, but they are not present anymore in the latest Magento version.

**Comment by jonatanrdsantos** [comment]
Hi @Davie82 are you sure that `stock_after_insert` and `stock_before_update` are Magento 2 core triggers and not from a customized module?

I'm saying that because in my care, the triggers remaining with a problem are those from customized modules!

**Comment by Davie82** [comment]
Hi @jonatanrdsantos no not quite sure. I couldn't find any reference in the external modules either. But maybe this will help someone with the same issue.

**Comment by engcom-November** [comment]
Hello @Nuranto,

It seems like on a clean installation, setup:upgrade works without any problem.
As pointed out here https://github.com/magento/magento2/issues/35735#issuecomment-1625438162, due to a custom or third party module's trigger we are getting these errors.
If you can mention the triggers or any third party module that is adding these triggers with which you are facing this issue, we might be able to reproduce it.

Thank you.

**Comment by Linek** [comment]
We had this error at the same day on Production and Preproduction deployment.
Repeating manually setup:upgrade was successful. 

The only change we were deploying was design changes in phtml. 

I believe this problem is not related with the change in the code, but more with infrastructure changes, but no idea for now what it could be.

**Comment by mpchadwick** [comment]
I believe this can happen when running setup:upgrade during deployment of code which adds a new mview. We are upgrading and had this crash. I can see this from an strace right before the crash

```
3SELECT `mview_state`.* FROM `mview_state` WHERE (`mview_state`.`view_id`='klevu_product_sync_stock')", 105,
 MSG_DONTWAIT, NULL, 0) = 105
```

However if I check `mview_state` there is not yet any view id `klevu_product_sync`

There reference to that is in `mview.xml`

```
<view id=\"klevu_product_sync_stock\" class=\"Klevu\\Search\\Model\\Indexer\\Sync\\ProductStockSyncIndexer\" gro
up=\"indexer\">\n        <subscriptions>\n            <table name=\"cataloginventory_stock_item\" entity_column=\"product_id\"
/>\n            <!--Track product status to trigger stock indexer-->\n            <table name=\"catalog_product_entity_int\" en
tity_column=\"entity_id\" />\n        </subscriptions>\n    </view>
```

Seems like during setup:upgrade it is referring to mview.xml for list of mviews and assuming they already exist in `mview_state` which they won't if the code is being deployed for the first time.

**Comment by Linek** [comment]
Hi @engcom-Hotel 
The problem guys is that this issue is not reproducible at all. Happened once - on prod and preprod deployment and never came back :)

**Comment by PhanQuocPhu** [comment]
Facing this issue when upgrading our site from magento 2.4.3-p1 to 2.4.6-p3.
It got stuck when removing unused trigger:
![image](https://github.com/magento/magento2/assets/60659261/6ed6f231-a35a-44f2-800d-ede92a5e6182)
In my case it was because of an third party module created an triggered in the 'cataloginventory_stock_item' table. This module go live after I start the upgrading ticket so my codebase doesn't contain the latest commit. When I used the latest production database with my current codebase, I will get this error when run: se:up
I drop the trigger and merge new commit into my codebase, and everything get back to work again.

**Comment by engcom-Hotel** [comment]
Thanks @Linek @PhanQuocPhu,

It seems the issue is not reproducible. Hence we are closing this issue.

Let us know if you are facing this again in the same issue, we will happy to help you.

Thanks

**Comment by Genaker** [comment]
issue still exists patch fixes it 
![image](https://github.com/magento/magento2/assets/9213670/d54e3e6b-b4bd-4ea5-8a6b-de75788d965e)

**Comment by raghuiamsingh** [comment]
will this ever be included in the release, so that it gets fixed for all of us?

**Comment by ejaz-alam** [comment]
getting the same error. :( any solution?

**Comment by Genaker** [comment]
> Getting the same error. :( any solution?

Check the solution provided by me. You removed some extensions and left some related attributes in the database. You need to delete them manually from the EAV attribute table. Magento 2 sucks!

**Comment by Genaker** [comment]
> the

Sorry, this one is a different issue. It is not related to the attribute. Sorry. User the patch

**Comment by elevinskii** [comment]
The patch is available in the 2.4.8-beta1.

https://github.com/magento/magento2/commit/88688710293bdec3de9bdc19df2419ce1ee7c7b4
